src/metabase/feature_extraction/automagic_dashboards.clj

```diff
@@ -0,0 +1,335 @@
+             [permissions :as perms]
+             [table :refer [Table]]]
+            [toucan
+             [db :as db]
+             [hydrate :refer [hydrate]]]
+            [yaml.core :as yaml]))
+
+(def ^:private ^Integer max-score 100)
+
+(defn- linked-tables
+  "Return all tables accessable from a given table and the paths to get there."
+  [table]
+  (map (fn [{:keys [id fk_target_field_id]}]
+         {:table (-> fk_target_field_id Field :table_id Table)
+          :fk    id})
+       (db/select [Field :id :fk_target_field_id]
+         :table_id (:id table)
+         :fk_target_field_id [:not= nil])))
+
+(defmulti
+  ^{:doc "Get a MBQL reference for a given model."
+    :arglists '([query-type model])
+    :private true}
+  ->reference (fn [query-type model]
+                [query-type (type model)]))
+
+(defmethod ->reference [:mbql (type Field)]
+  [_ {:keys [fk_target_field_id id link]}]
+  (cond
+    link               [:fk-> link id]
+    fk_target_field_id [:fk-> id fk_target_field_id]
+    :else              [:field-id id]))
+
+(defmethod ->reference [:mbql clojure.lang.PersistentVector]
+  [_ form]
+  form)
+
+(defn- ->type
+  [x]
+  (if (keyword? x)
+    x
+    (keyword "type" x)))
+
+(defn- filter-fields
+  [fieldspec table]
+  (filter (fn [{:keys [base_type special_type]}]
+            (or (isa? base_type fieldspec)
+                (isa? special_type fieldspec)))
+          (db/select Field :table_id (:id table))))
+
+(defn- field-candidates
+  ([context fieldspec]
+   (let [fieldspec (->type fieldspec)]
+     (filter-fields fieldspec (:root-table context))))
+  ([context tablespec fieldspec]
+   (let [fieldspec            (->type fieldspec)
+         tablespec            (->type tablespec)
+         [{:keys [table fk]}] (->> context
+                                 :linked-tables
+                                 (filter #(-> %
+                                              :table
+                                              :entity_type
+                                              (isa? tablespec))))]
+     (some->> table
+              (filter-fields fieldspec)
+              (map #(assoc % :link fk))))))
+
+(defn- op?
+  [form]
+  (and (sequential? form)
+       ((some-fn keyword? string?) (first form))))
+
+(defmulti
+  ^{:doc ""
+    :arglists '([context [op & args]])
+    :private true}
+  op (fn [_ [op & _]]
+       (if (keyword? op)
+         op
+         (keyword (s/lower-case op)))))
+
+(defmethod op :field-type
+  [context [_ & typespec :as form]]
+  {form (apply field-candidates context typespec)})
+
+(defmethod op :dimension
+  [context [_ dimension :as form]]
+  {form (-> context :dimensions (get dimension) :matches)})
+
+(defmethod op :default
+  [_ _]
+  nil)
+
+(defn- bindings-candidates
+  [context form]
+  (->> form
+       (tree-seq (some-fn map? vector?) identity)
+       (filter op?)
+       (keep (partial op context))
+       (apply merge)))
+
+(defn- form-candidates
```

Comment:
what's a form? This is never defined and the term isn't used anywhere else in the codebase.